api/src/test/java/org/openmrs/api/OrderServiceTest.java

```diff
@@ -2969,4 +2969,178 @@ public void saveOrder_shouldPassIfAnKnownDrugOrderForTheSameDrugFormulationSpeci
+		drugOrder.setQuantity(duplicateOrder.getQuantity());
+		drugOrder.setQuantityUnits(duplicateOrder.getQuantityUnits());
+		drugOrder.setNumRefills(duplicateOrder.getNumRefills());
+
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage("Cannot have more than one active order for the same orderable and care setting");
+		orderService.saveOrder(drugOrder, null);
+	}
+
+	@Test
+	@Verifies(value = "should discontinue previousNonCodedOrder if it is not already discontinued", method = "saveOrder(Order)")
+	public void saveOrder_shouldDiscontinuePreviousNonCodedOrderIfItIsNotAlreadyDiscontinued() throws Exception {
+		//We are trying to discontinue order id 584 in OrderServiceTest-nonCodedDrugs.xml
+		executeDataSet("org/openmrs/api/include/OrderServiceTest-nonCodedDrugs.xml");
+		DrugOrder previousOrder = (DrugOrder) orderService.getOrder(584);
+		DrugOrder drugOrder = previousOrder.cloneForDiscontinuing();
+		drugOrder.setPreviousOrder(previousOrder);
+		drugOrder.setDateActivated(new Date());
+		drugOrder.setOrderer(providerService.getProvider(1));
+		drugOrder.setEncounter(encounterService.getEncounter(6));
+
+
+		Order saveOrder = orderService.saveOrder(drugOrder, null);
+		Assert.assertNotNull("previous order should be discontinued", previousOrder.getDateStopped());
+		assertNotNull(orderService.getOrder(saveOrder.getOrderId()));
+	}
+
+	@Test
+	@Verifies(value = "should fail discontinue previousNonCodedDrugOrder if the orderable of the previous order and the new one order don't match ", method = "saveOrder(Order)")
+	public void saveOrder_shouldFailDiscontinueNonCodedDrugOrderIfOrderableOfPreviousAndNewOrderDontMatch() throws Exception {
+		executeDataSet("org/openmrs/api/include/OrderServiceTest-nonCodedDrugs.xml");
+		DrugOrder previousOrder = (DrugOrder) orderService.getOrder(584);
+		DrugOrder drugOrder = previousOrder.cloneForDiscontinuing();
+		drugOrder.setDrugNonCoded("non coded drug paracetemol");
+		drugOrder.setPreviousOrder(previousOrder);
+		drugOrder.setDateActivated(new Date());
+		drugOrder.setOrderer(providerService.getProvider(1));
+		drugOrder.setEncounter(encounterService.getEncounter(6));
+
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage("The orderable of the previous order and the new one order don't match");
+		orderService.saveOrder(drugOrder, null);
+	}
+
+
+	@Test
+	@Verifies(value = "should fail revising previousNonCodedDrugOrder if the orderable of the previous order and the new one order don't match ", method = "saveOrder(Order)")
+	public void saveOrder_shouldFailIfDrugNonCodedInPreviousDrugOrderDoesNotMatchThatOfTheRevisedDrugOrder() throws Exception {
+		executeDataSet("org/openmrs/api/include/OrderServiceTest-nonCodedDrugs.xml");
+		DrugOrder previousOrder = (DrugOrder)orderService.getOrder(584);
+		DrugOrder order = previousOrder.cloneForRevision();
+		String drugNonCodedParacetemol = "non coded paracetemol";
```

Comment:
Would be nice to have an asssertion that this isn't equal to the previous' non coded value